File: src/python/ko_pii/detect.py
"""All-detector entry point — 모든 카테고리 검출기를 한 번에 실행."""
from __future__ import annotations

import re
from typing import Iterable, Optional

from ko_pii.core.types import DetectionResult
from ko_pii.core.overlap import resolve_overlaps as _resolve_overlaps
from ko_pii.core.unicode_norm import (
    needs_normalization,
    normalize_unicode,
    remap_to_source,
)
from ko_pii.domain import civil_petition as _dom_petition
from ko_pii.domain import government as _dom_gov
from ko_pii.domain import hr as _dom_hr
from ko_pii.patterns import (
    account,
    address,
    birth,
    business_reg,
    card,
    corp_reg,
    court_case,
    driver_license,
    edi_drug,
    email,
    fax,
    frn,
    ip,
    medical_insurance,
    nationality,
    passport,
    person,
    personal_attr,
    phone,
    pnu,
    postal_code,
    prescription,
    rrn,
    url,
    vehicle,
)

DETECTORS = (
    rrn.detect,
    frn.detect,
    business_reg.detect,
    corp_reg.detect,
    driver_license.detect,
    passport.detect,
    card.detect,
    medical_insurance.detect,
    prescription.detect,
    pnu.detect,
    fax.detect,
    phone.detect,
    email.detect,
    postal_code.detect,
    ip.detect,
    vehicle.detect,
    url.detect,
    address.detect,
    nationality.detect,
    account.detect,
    person.detect,
    # 식의약·법조 도메인
    edi_drug.detect,
    court_case.detect,
    # 인적 속성 (준식별자) — 학력·전공·직책·측정치 + 생년월일
    birth.detect,
    personal_attr.detect,
    # Domain-specific
    _dom_gov.detect,
    _dom_petition.detect,
    _dom_hr.detect,
)

# 영숫자 런(ASCII + 전각). 원본 재검사 필요 여부 판정용 — invisible 제거가 이 '런의 모양'
# (길이 시퀀스)을 바꿨다면 두 PII 가 경계 융합/분리됐다는 뜻이라 원본도 봐야 한다.
_PII_RUN = re.compile(r"[0-9A-Za-z０-９Ａ-Ｚａ-ｚ]+")


def _run_shape(s: str) -> list[int]:
    return [m.end() - m.start() for m in _PII_RUN.finditer(s)]


def detect_all(
    text: str,
    include: Optional[Iterable[str]] = None,
    exclude: Optional[Iterable[str]] = None,
    *,
    normalize: bool = True,
) -> list[DetectionResult]:
    """Run every detector and return a merged, conflict-resolved list.

    Overlapping spans are resolved by priority — higher risk level, then higher
    confidence, then longer span, then earlier start — so a lower-priority span
    never shadows a higher-confidence PII span (see ``_resolve_overlaps``).

    ``include`` / ``exclude`` filter on the resulting DetectionResult labels.

    ``normalize`` (기본 True): 전각/호환문자 폴딩 + 제로폭 문자 제거로 검출
    우회를 차단한다. 결과 offset 은 원본 ``text`` 기준으로 역매핑된다.
    ASCII 입력은 우회 벡터가 없어 그대로 통과한다.
    """
    if not isinstance(text, str):
        raise TypeError(f"detect_all() expects str, got {type(text).__name__}")
    source = text
    offset_map: Optional[list[int]] = None
    if normalize and needs_normalization(text):
        norm, omap = normalize_unicode(text)
        if norm != text:
            text, offset_map = norm, omap

    raw: list[DetectionResult] = []
    for fn in DETECTORS:
        raw.extend(fn(text))

    if offset_map is not None:
        # 정규화가 텍스트를 바꿨다(전각 폴딩/보이지 않는 문자 제거 등).
        # 정규화본 검출 offset 을 원본으로 역매핑한 뒤, **원본에도 검출을 한 번 더
        # 수행해 합집합**한다. 보이지 않는 문자(ZWSP·소프트하이픈 등) 제거가 인접
        # PII 두 개(예: RRN​전화)를 하나의 숫자열로 융합시켜 양쪽 검출기의 경계
        # 가드가 둘 다 거부하던 누출을 막는다 — 원본에서는 그 문자가 경계 역할을
        # 해 둘 다 검출된다. 정상 PII 중복은 아래 겹침 해소가 합쳐준다.
        raw = remap_to_source(raw, offset_map, source)
        # 원본 재검사는 invisible 제거가 '영숫자 런 모양'을 바꾼 경우(=PII 경계 융합/분리)만
        # 한다. 한글/공백/이모지 사이의 제로폭 떡칠은 모양이 그대로라 융합이 없으므로
        # 더블패스(검출기 28개 2회)를 건너뛴다 — 제로폭 1자로 비용을 2배 만드는 DoS 완화.
        if _run_shape(source) != _run_shape(text):
            for fn in DETECTORS:
                raw.extend(fn(source))

    inc = set(include) if include else None
    exc = set(exclude) if exclude else set()
    if inc is not None:
        raw = [d for d in raw if d.label in inc]
    if exc:
        raw = [d for d in raw if d.label not in exc]

    return _resolve_overlaps(raw)


# 겹침 해소는 core.overlap.resolve_overlaps 단일 구현 사용 (위 import 의 _resolve_overlaps).
# detect_all / merge_detections(integrations.hybrid) / 치환(modes._apply) 이 모두 같은
# 우선순위(위험도→확신도→길이)로 해소해 누출 회귀를 차단한다.

File: src/ts/src/detect.ts
/**
 * 전체 검출기 진입점 — Python ko_pii.detect 대응.
 *
 * 모든 검출기를 직렬 실행하고 정규화 원본 재검사 합집합 + 겹침 해소로 병합한다.
 */

import { resolveOverlaps } from "./core/overlap.js";
import type { DetectionResult } from "./core/types.js";
import { needsNormalization, normalizeUnicode, remapToSource } from "./core/unicodeNorm.js";
import * as civilPetition from "./domain/civil_petition.js";
import * as government from "./domain/government.js";
import * as hr from "./domain/hr.js";
import * as account from "./patterns/account.js";
import * as address from "./patterns/address.js";
import * as birth from "./patterns/birth.js";
import * as businessReg from "./patterns/business_reg.js";
import * as card from "./patterns/card.js";
import * as corpReg from "./patterns/corp_reg.js";
import * as courtCase from "./patterns/court_case.js";
import * as driverLicense from "./patterns/driver_license.js";
import * as ediDrug from "./patterns/edi_drug.js";
import * as email from "./patterns/email.js";
import * as fax from "./patterns/fax.js";
import * as frn from "./patterns/frn.js";
import * as ip from "./patterns/ip.js";
import * as medicalInsurance from "./patterns/medical_insurance.js";
import * as nationality from "./patterns/nationality.js";
import * as passport from "./patterns/passport.js";
import * as person from "./patterns/person.js";
import * as personalAttr from "./patterns/personal_attr.js";
import * as phone from "./patterns/phone.js";
import * as pnu from "./patterns/pnu.js";
import * as postalCode from "./patterns/postal_code.js";
import * as prescription from "./patterns/prescription.js";
import * as rrn from "./patterns/rrn.js";
import * as url from "./patterns/url.js";
import * as vehicle from "./patterns/vehicle.js";

/** Python 검출기는 Iterator 를 반환하므로 Iterable 로 수용한다. */
export type Detector = (text: string) => Iterable<DetectionResult>;

/** 실행 순서는 Python detect.py 의 DETECTORS 와 동일하게 유지한다. */
export const DETECTORS: readonly Detector[] = [
  rrn.detect,
  frn.detect,
  businessReg.detect,
  corpReg.detect,
  driverLicense.detect,
  passport.detect,
  card.detect,
  medicalInsurance.detect,
  prescription.detect,
  pnu.detect,
  fax.detect,
  phone.detect,
  email.detect,
  postalCode.detect,
  ip.detect,
  vehicle.detect,
  url.detect,
  address.detect,
  nationality.detect,
  account.detect,
  person.detect,
  // 식의약·법조 도메인
  ediDrug.detect,
  courtCase.detect,
  // 인적 속성 (준식별자) — 학력·전공·직책·측정치 + 생년월일
  birth.detect,
  personalAttr.detect,
  // Domain-specific
  government.detect,
  civilPetition.detect,
  hr.detect,
];

// 영숫자 런(ASCII + 전각) — 원본 재검사 필요 여부 판정용.
const PII_RUN = /[0-9A-Za-z０-９Ａ-Ｚａ-ｚ]+/g;

function runShape(s: string): number[] {
  PII_RUN.lastIndex = 0;
  return [...s.matchAll(PII_RUN)].map((m) => m[0].length);
}

export interface DetectAllOptions {
  include?: Iterable<string>;
  exclude?: Iterable<string>;
  /** 전각/호환문자 폴딩 + 제로폭 제거 (기본 true) — 결과 offset 은 원본 기준으로 역매핑. */
  normalize?: boolean;
}

export function detectAll(
  text: string,
  include?: Iterable<string> | null,
  exclude?: Iterable<string> | null,
  options?: { normalize?: boolean },
): DetectionResult[] {
  const normalize = options?.normalize ?? true;
  if (typeof text !== "string") {
    throw new TypeError(`detectAll() expects string, got ${typeof text}`);
  }
  const source = text;
  let offsetMap: number[] | null = null;
  let work = text;
  if (normalize && needsNormalization(text)) {
    const [norm, omap] = normalizeUnicode(text);
    if (norm !== text) {
      work = norm;
      offsetMap = omap;
    }
  }

  const raw: DetectionResult[] = [];
  for (const fn of DETECTORS) {
    raw.push(...fn(work));
  }

  if (offsetMap !== null) {
    // 정규화가 텍스트를 바꿨다: offset 을 원본으로 역매핑한 뒤 원본에도 한 번 더
    // 검출해 합집합(누출 차단). 원본 재검사는 영숫자 런 모양이 바뀐 경우만 — DoS 완화.
    const remapped = remapToSource(raw, offsetMap, source);
    if (JSON.stringify(runShape(source)) !== JSON.stringify(runShape(work))) {
      for (const fn of DETECTORS) {
        remapped.push(...fn(source));
      }
    }
    raw.length = 0;
    raw.push(...remapped);
  }

  // Python: `set(include) if include else None` — 빈 include 는 "필터 없음"이고,
  // include 와 exclude 는 둘 다 적용된다.
  const incSet = include ? new Set(include) : null;
  const inc = incSet !== null && incSet.size > 0 ? incSet : null;
  const exc = exclude ? new Set(exclude) : new Set<string>();
  let filtered = raw;
  if (inc !== null) filtered = filtered.filter((d) => inc.has(d.label));
  if (exc.size > 0) filtered = filtered.filter((d) => !exc.has(d.label));

  return resolveOverlaps(filtered);
}
